### DISCUSSION TITLE
Weird behaviour in the log

### DISCUSSION THREAD

**Discussion by KestrelKitezh** [streamlink/streamlink#4438] [Q&A] [answered]
So I've been using streamlink for a long while, on several computers in my house. One of the computers give a weird vlc error each time I start a twitch stream though, and I just wondered if anyone knew why. It doesn't really matter since he starts the stream regardless, but I guess I'm just a curious guy.

- Found matching plugin twitch for URL twitch.tv/fairlight_excalibur
- Available streams: audio_only, 160p (worst), 360p, 480p, 720p60, 1080p60 (best)
- Opening stream: 1080p60 (hls)
- Starting player: C:\Program Files\VideoLAN\VLC\vlc.exe
- [000002bd2d9b3590] main libvlc: Running vlc with the default interface. Use 'cvlc' to use vlc without interface.
- [000002bd2daec140] filesystem stream error: cannot open file H\e\r\o\e\s\ (No such file or directory)
- [000002bd2daebcc0] filesystem stream error: cannot open file o\f\ (No such file or directory)
- [000002bd2d9b3590] main libvlc: Status file authenticated
- [000002bd2daec140] filesystem stream error: cannot open file H\a\m\m\e\r\w\a\t\c\h (No such file or directory)
- [000002bd30323950] mpeg4audio packetizer: AAC channels: 2 samplerate: 48000
- [000002bd314fcc90] avcodec decoder: Using D3D11VA (NVIDIA GeForce GTX 1080 Ti, vendor 10de(NVIDIA), device 1b06, revision a1) for hardware decoding
- [000002bd2d9b3890] main playlist: end of playlist, exiting
- [000002bd30d34b60] direct3d11 vout display error: SetThumbNailClip failed: 0x800706f4

Why is he attempting to create all those sub folders? The recent folder in my explorer is cluttered with hundreds of those single "digit" sub folders which doesn't exist, and it doesn't really make any sense (in my mind). I just reinstalled streamlink, twitchgui just in case, and reinstalled vlc as well, but still this happens just on this computer. Is it a file access issue?

![explorer_screenshot](https://user-images.githubusercontent.com/44786999/162430620-ea21273d-63c2-479f-b945-9e48d96b03a8.png)

**Comment by bastimeyer** [comment, marked as answer]
> * [000002bd2daec140] filesystem stream error: cannot open file H\e\r\o\e\s\ (No such file or directory)
> * [000002bd2daebcc0] filesystem stream error: cannot open file o\f\ (No such file or directory)
> * [000002bd2d9b3590] main libvlc: Status file authenticated
> * [000002bd2daec140] filesystem stream error: cannot open file H\a\m\m\e\r\w\a\t\c\h (No such file or directory)

This very much looks like you've misconfigured Streamlink's launch parameters in Streamlink Twitch GUI in regards to the player title and custom variables, and you didn't wrap custom variables in quotes. This then leads to a broken [`--player`](https://streamlink.github.io/cli.html#cmdoption-player) argument value when launching Streamlink, depending on the stream's title which was substituted by Streamlink Twitch GUI.

**Post the entire command line that gets generated by setting the `--loglevel=debug` parameter**, either as a parameter when launching Streamlink Twitch GUI, so that it logs how it invokes the streamlink command with its parameters, or set the same parameter as a Streamlink launch parameter, so it shows which arguments were set in its debug log header.

- https://github.com/streamlink/streamlink-twitch-gui/wiki/Parameters#--loglevel--l
- https://streamlink.github.io/cli.html#cmdoption-loglevel


By the way, Streamlink Twitch GUI explicitly warns you in the GUI itself as well as its wiki that you must set any custom variables into quotes:

- https://github.com/streamlink/streamlink-twitch-gui/wiki/Player-configuration#custom-player-parameters
- https://github.com/streamlink/streamlink-twitch-gui/blob/v2.0.0/src/app/locales/en/components.yml#L65-L69

----

The value of Streamlink's `--player` argument gets executed as a whole, so if you are not careful, then this can lead to accidental code execution if the value includes substituted strings like stream titles. Streamlink Twitch GUI already tries its best to avoid this by escaping every single substituted character if the variable that gets substituted was not wrapped by the user in a quoted string (according to its tokenization logic), so that Streamlink can't accidentally execute stuff when launching the player, but this can't prevent the issue entirely.

The title of the stream you're trying to watch is "RETURN TO HAMMERWATCH", so I guess it must've something to do with the "RETURN" keyword which then gets interpreted as a shell command, which means you must have misconfigured the `--player` value. Once again, post the entire output after setting the `--loglevel=debug` parameter.

On Windows, the way Streamlink handles its `--player` CLI argument is especially a problem with how Python's `subprocess.Popen` works and that the string does not get parsed via `shlex.split`:

- https://github.com/streamlink/streamlink/blob/3.2.0/src/streamlink_cli/output.py#L196-L199
- https://github.com/streamlink/streamlink/blob/3.2.0/src/streamlink_cli/output.py#L231-L241
- https://docs.python.org/3/library/subprocess.html#subprocess.Popen
- https://docs.python.org/3/library/shlex.html

I never liked the `--player` argument implementation in Streamlink. This comes from the Livestreamer days before the fork, and it's a result of laziness and carelessness, which is dangerous, as can be seen in this case. The `--player-args` argument already exists, so it's not necessary to interpret the `--player` argument value as a shell command, and it should only be used as a path to the player executable. Users should always use the `--player-args` argument for setting custom player args.

The `--player` argument implementation was never changed in Streamlink because changing it would be a major breaking change and would guaranteed cause lots of trouble for people who don't know what they're doing and who don't read changelogs. I'm very much in favor of fixing it though and relying on `--player-args` for any custom player launch arguments.

Streamlink Twitch GUI has been built with this exact usage in mind since the beginning, but as a launcher it also allows the user to use different configurations.

**Reply by KestrelKitezh** [reply]
> This very much looks like you've misconfigured Streamlink's launch parameters in Streamlink Twitch GUI in regards to the player title and custom variables, and you didn't wrap custom variables in quotes. 

This is out of the box. I haven't modified anything in either streamlink config, vlc or the gui. This happened even after deleting all the config files and reinstalling all 3 aswell. Adding the debug soon

**Reply by bastimeyer** [reply]
Post the `--loglevel=debug` output

**Reply by KestrelKitezh** [reply]
[2022-04-08_14-28-42.log](https://github.com/streamlink/streamlink/files/8452717/2022-04-08_14-28-42.log)

**Reply by bastimeyer** [reply]
```
[2022-04-08T14:28:48.178Z][debug][StreamingService]
Resolving player
{
    "player": "vlc",
    "playerUserData": {
        "instance": true,
        "close": true,
        "title": true,
        "minimal": true,
        "exec": "C:\\Program Files\\VideoLAN\\VLC\\vlc.exe",
        "args": "{game}",
        "type": "settings-streaming-player-vlc"
    }
}

[2022-04-08T14:28:48.181Z][debug][StreamingService]
Resolved player
{
    "exec": "C:\\Program Files\\VideoLAN\\VLC\\vlc.exe",
    "params": "{game} --no-one-instance --play-and-exit --input-title-format \"{name} - {game} - {status}\" --qt-minimal-view",
    "env": null
}

[2022-04-08T14:28:48.207Z][debug][StreamingService]
Spawning process
{
    "exec": "C:\\Program Files (x86)\\Streamlink\\bin\\streamlink.exe",
    "params": [
        "--twitch-disable-hosting",
        "--twitch-disable-ads",
        "--twitch-low-latency",
        "--player",
        "C:\\Program Files\\VideoLAN\\VLC\\vlc.exe",
        "--player-args",
        "\\D\\o\\t\\a\\ \\2 --no-one-instance --play-and-exit --input-title-format \"Gorgc - Dota 2 - >.<\" --qt-minimal-view",
        "--player-passthrough",
        "hls",
        "--hls-live-edge",
        "3",
        "--stream-segment-threads",
        "1",
        "--retry-open",
        "3",
        "--retry-streams",
        "1",
        "twitch.tv/gorgc",
        "best"
    ],
    "env": null
}
```

>         "exec": "C:\\Program Files\\VideoLAN\\VLC\\vlc.exe",
>         "args": "{game}",

You have set the player args as `{game}` for some reason. Remove it. This doesn't belong there and will get interpreted by your player as additional input parameter(s). Since you didn't quote it as well, VLC will try to open `\D\o\t\a\ \2`.

You've also selected the HLS passthrough option in the Twitch GUI settings menu (hence the `--player-passthrough=hls` Streamlink parameter), which completely bypasses Streamlink and just passes the resolved HLS stream URL to VLC. This should only be done if you know exactly what you're doing.

**Reply by KestrelKitezh** [reply]
So this would mean some file eluded me when I reinstalled the gui then?

Thanks for the quick answer. Love streamlink so much love <3

**Reply by bastimeyer** [reply]
Streamlink Twitch GUI's configuration is stored in the application's "user data directory", which is a Chromium thing:
https://github.com/streamlink/streamlink-twitch-gui/wiki/Parameters#user-data-directory

Streamlink's separate config file is stored here:
https://streamlink.github.io/latest/cli/config.html#location
